### ISSUE TITLE
Vim bindings

### ISSUE THREAD

**Issue by felipesere** [marktext/marktext#596]
### Describe your feature request

It would be lovely to have sane subset of vim bindings (navigation, and a few of the motions) available in MarkText. 

### What problem does this feature solve? [optional]

Editing text in a way that I am used to. It makes the transition from `vim` to `MarkText` for writing much smoother.

### Additional context [optional]

I understand this opens the door for Emacs and other bindings.

**Comment by mlopezgva** [comment]
If MarkText uses it, ACE editor already have this feature, and it is very well implemented, by the way. I use it in Cloud9 (place of birth of ACE) and it almost feels like gvim. It can even fold using `zc` `zo`, and a very limited command-line mode (in c9, it supports `:w` and a few).

It would be nice to have it. And, yes, ACE accepts Emacs and sublime modes (including multi-cursors).

**Comment by fxha** [comment]
@Yihen Sorry, it's currently not planned.

**Comment by spetex** [comment]
@fxha Hello, I am interested in helping with this feature, how can I get started? Are there any new contributors guideline?

**Comment by Jocs** [comment]
@spetex FYI https://github.com/marktext/marktext/blob/develop/.github/CONTRIBUTING.md

if you have any question you can ask me or @fxha

**Comment by fxha** [comment]
@spetex Reading the contribution guideline is a good starting point. Unfortunately, we currently don't have good documentation but I think you should implement a VIM class that use a state machine to track input and add it to the `Muya` class. This VIM instance should controls all input like `muya/lib/eventHandler/keyboard.js` methods and allow input depending on the state. Features like `:s` may be as problem because we don't have an editor API yet, only event listeners. Also the communication between fontend and backend need optimizations.

**Comment by mwesterhof** [comment]
Just to chime in an fxha's comment;

As a self-confessed vim-lover, I'd be absolutely thrilled to see even a very basic support for modal editing added to marktext. Having a full feature set would be amazing, but even just having the ability to switch between normal/insert/[visual] modes, and some basic motions, would be a game changer for me. I know I can't speak for everyone else, but i'm sorry to say it's the only thing holding me back from doing all my markdown writing using marktext. I would love to make that switch, though.

**Comment by spetex** [comment]
Yes, that is my plan. I am starting with just a small feature set to get going. Although it is going to take me some time as I am now trying to read through the code and learn how it works first.

**Comment by brandonkal** [comment]
Seeing that this is a JavaScript project, those who are looking to add this feature should have a look at https://github.com/akahuku/wasavi as that is a JavaScript only vim implementation.

Another approach is to integrate with nvim.

**Comment by vijay609** [comment]
@spetex Were you able to look into this?
It would be great to get VIM keybindings working. I would love to help.

**Comment by spetex** [comment]
@vijay609 Unfortunately not. I have moved to a different rich text editor based on SlateJS.

**Comment by cmsax** [comment]
Why not just use Vim with markdown plugins installed as an editor and use MarkText as a viewer/editor to do further work?

**Comment by Shatonly** [comment]
Sadly, I ended up using only vim with a preview in web browser, and pandoc to create pdfs, but this was a difficult choice.

**Comment by fazlearefin** [comment]
vim keybindings is the only reason why I am still using Obsidian. vim keybindings in marktext would be awesome

**Comment by gloryVine** [comment]
I'm the same situation as @fazlearefin. I'm sticking to Obsidian purely because of the vim keybindings, and would switch over to marktext immediately if it had these.

**Comment by gfauredev** [comment]
Roughly the same, I still use vim for everything due to its keybindings, but it would be so nice to have live in-line markdown rendering, pictures …

**Comment by sisrfeng** [comment]
Obsidian with its vimrc plugin is not bad, but not complete vim.

I give up waiting for typora' s vim-keybind
https://github.com/typora/typora-issues/issues/187

**Comment by ayhon** [comment]
Just dropping in a comment to show support for this issue. Would love some sort of vim-like keybindings.

**Comment by ATNewHope** [comment]
Just dropping in a comment to show support for this issue. Would love some sort of vim-like keybindings.
Btw, Zettlr, another opensource MD editor support both vim and emacs bindings, however, I don't like some of its designs compared to marktext. So I still wish the vim binding on marktext could be added, that would make marktext perfect for me.

**Comment by godot42x** [comment]
This is a 2016 issue:
https://github.com/typora/typora-issues/issues/187

It would be a big step if MarkText supported VIM embaled, or if plug-in modules were enabled.

**Comment by AhmedHalim96** [comment]
It would be amazing if we could have vim mode in marktext. it would make marktext more awesome.

**Comment by nivethan-me** [comment]
Adding a comment to show support for this issue. love to see even a basic navigation vim keybindings with normal and insert mode.

**Comment by asolopovas** [comment]
I am in support of this feature also, when you learn vim you just understand how everythig else is inferior in terms of speed when working with computer (at least for now before SKYNET takes over). I think features need some sort of money pool to be implemented so that anyone can contribute money to this pool in any currency they prefer including crypto. This way everyone can vote for features with their dollars and thus benefit from them and incentivise developers to implement them, and then do a refund to everyone in set period of time (3, 6, 12 months) depending on the complexity of the task that can be determined if features don't get implemented...

Its a bit of a gamble due to inflation but I am willing to bet a little money to improve the experience I have with computers, my time is more precious than my money. It is especially cool if it would help to grow opensource community around the globe.

**Comment by TomasTillmann** [comment]
I would love to use Vim keybindings in MarkText. Is someone or has someone worked on it? Looking at the PRs and based on this issues activity it looks like nobody has started.

 So I guess I have to start from scratch. If someone would like to help, contact me please, would definitely appreciate it.

**Comment by asolopovas** [comment]
Maybe with help of AI this can happen finally?

**Comment by Jocs** [comment]
This has been the most requested feature, and we'll be adding it in version 0.20.0.
